plugins/com.google.cloud.tools.eclipse.appengine.standard.java8/src/com/google/cloud/tools/eclipse/appengine/standard/java8/AppEngineWebBuilder.java

```diff
@@ -0,0 +1,104 @@
+import com.google.cloud.tools.eclipse.appengine.facets.WebProjectUtil;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResourceDelta;
+import org.eclipse.core.resources.IncrementalProjectBuilder;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.jst.common.project.facet.core.JavaFacet;
+import org.eclipse.jst.j2ee.web.project.facet.WebFacetUtils;
+import org.eclipse.wst.common.project.facet.core.IFacetedProject;
+import org.eclipse.wst.common.project.facet.core.IFacetedProject.Action;
+import org.eclipse.wst.common.project.facet.core.ProjectFacetsManager;
+
+/**
+ * Builder to monitor user changes relating to the {@code <runtime>java8</runtime>} element in
+ * {@code appengine-web.xml}, and to reflect such changes into the project facets.
+ */
+public class AppEngineWebBuilder extends IncrementalProjectBuilder {
+  private static final Logger logger = Logger.getLogger(AppEngineWebBuilder.class.getName());
+  static final String BUILDER_ID =
+      "com.google.cloud.tools.eclipse.appengine.standard.java8.appengineWeb";
+
+  @Override
+  protected IProject[] build(int kind, Map<String, String> args, IProgressMonitor monitor)
+      throws CoreException {
+    IFacetedProject project = ProjectFacetsManager.create(getProject());
+    if (project == null || !AppEngineStandardFacet.hasFacet(project)) {
+      logger.fine(getProject() + ": no build required: no App Engine Standard facet");
+      return null;
+    }
+    IFile appEngineWebDescriptor =
+        WebProjectUtil.findInWebInf(project.getProject(), new Path("appengine-web.xml"));
+    if (appEngineWebDescriptor == null || !appEngineWebDescriptor.exists()) {
+      logger.warning(getProject() + ": no build required: missing appengine-web.xml");
+      return null;
+    }
+    if (kind == FULL_BUILD) {
+      checkRuntimeElement(project, appEngineWebDescriptor, monitor);
+    } else {
+      IResourceDelta delta = getDelta(getProject());
+      if (delta == null
+          || delta.findMember(appEngineWebDescriptor.getProjectRelativePath()) != null) {
+        checkRuntimeElement(project, appEngineWebDescriptor, monitor);
+      } else {
+        logger.finer(getProject() + ": no build required: appengine-web.xml not changed");
+      }
+    }
+    return null;
+  }
+
+  private void checkRuntimeElement(IFacetedProject project, IFile appEngineWebDescriptor,
+      IProgressMonitor monitor) {
+    try (InputStream input = appEngineWebDescriptor.getContents()) {
+      boolean isJava8 = AppEngineDescriptor.parse(input).isJava8();
+      if (project.hasProjectFacet(JavaFacet.VERSION_1_8) != isJava8) {
```

Comment:
boolean != boolean 
is a bit surprising. I wonder if it clarify to refactor as
if 
else if
else